README.md
# Q Applet: RAM Usage

Displays the RAM usage on a Das Keyboard Q Series.

[GitHub repository](https://github.com/daskeyboard/daskeyboard-applet--ram-usage)

## Example

Displays the percentage of memory used with the following colors: greens &lt; 50%, yellow &lt; 70%, orange &lt; 85% and red 85% and above.

Example with 50% memory used :

![RAM Usage on a Das Keybaord Q](assets/image.png "Q RAM usage")

## Changelog

[CHANGELOG.MD](CHANGELOG.md)

## Installation

Requires a Das Keyboard Q series: www.daskeyboard.com

Installation, configuration and uninstallation of applets is done within
the Q Desktop application (https://www.daskeyboard.com/q )

## Running tests

- `yarn test`

## Contributions

Pull requests welcome.

## Copyright / License

Copyright 2014 - 2019 Das Keyboard / Metadot Corp.

Licensed under the GNU General Public License Version 2.0 (or later);
you may not use this work except in compliance with the License.
You may obtain a copy of the License in the LICENSE file, or at:

   http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Changelog Q Applet RAM Usage

## Version 1.0.0 - December 06, 2018

First release


test/test.js
const assert = require('assert');
const t = require('../index');

const black = '#000000';

describe('RamUsage', function () {
  let app = new t.RamUsage();

  it('#getRamUsage', function () {
    app.getRamUsage().then(percent => {
      console.log("RAM Usage is: " + percent);
      assert.ok(percent);
    })
  });


  it('#generateColor', function () {
    const testLight = function (percent) {
      const points = app.generateColor(percent);
      return points[0].color;
    }

    const assertLight = function (percent, expected) {
      const actual = testLight(percent);
      const message = `For RAM usage ${percent}, I have a ${actual} light `
        + `(expected ${expected})`;  
      assert(expected === actual, message);
    }

    assertLight(20, '#00FF00');
    assertLight(30, '#33FF00');
    assertLight(69, '#FFFF00');
    assertLight(73, '#FF6600');
    assertLight(84, '#FF6600');
    assertLight(90, '#FF0000');
    assertLight(100, '#FF0000');


  });

  it('#run()', function () {
    app.run().then((signal) => {
      console.log(JSON.stringify(signal));
      assert.ok(signal);
    }).catch(error => {
      assert.fail(error);
    });
  });
})

q-ram-usage.js
// Library to track ramUsage
const si = require('systeminformation');

// Library to send signal to Q keyboards
const q = require('daskeyboard-applet');

// Color associated to ram activity from low (green), middle (yellow), to high (red).
const colors = ['#00FF00', '#33FF00','#FFFF00', '#FF6600', '#FF0000'];

const logger = q.logger;


class RamUsage extends q.DesktopApp {
  constructor() {
    super();
    this.pollingInterval = 3000; // run() every 3000 ms
    logger.info("RAM Usage Meter ready to go!");
  }

  // call this function every pollingInterval
  async run() {
    return this.getRamUsage().then(percent => {
      // return signal each 3000ms, it depends of percent value
      return new q.Signal({
        points: [this.generateColor(percent)],
        name: "RAM Usage",
        message: "Memory used:" + Math.round(percent) + "%",
        isMuted: true, // don't flash the Q button on each signal
      });
    });
  }

  async getRamUsage() {
    return new Promise((resolve) => {
      si.mem(cb => {
       // get memory and create percent
       var percent = (cb.active/cb.total)*100;
       resolve(percent);
      });

 
    })
  }

  generateColor(percent) {

    let color =[];

    switch (true){

      case percent < 30:
        // return first color
        color.push(new q.Point(colors[0]));
        break;

      case percent < 50:
        // return second color
        color.push(new q.Point(colors[1]));
        break;

      case percent < 70:
        // return third color
        color.push(new q.Point(colors[2]));
        break;
      
      case percent < 85:
        // return fourth color
        color.push(new q.Point(colors[3]));
        break;

      case percent <= 100:
        // return fifth color
        color.push(new q.Point(colors[4]));
        break;

      default:
        // Something wrong happened, percent>100, return white color
        color.push(new q.Point("#FFFFFF"));
        break;

    };


    return color;
  }

}

module.exports = {
  RamUsage: RamUsage
};

const ramUsage = new RamUsage();


index.js
// Library to track ramUsage
const si = require('systeminformation');

// Library to send signal to Q keyboards
const q = require('daskeyboard-applet');

// Color associated to ram activity from low (green), middle (yellow), to high (red).
const colors = ['#00FF00', '#33FF00','#FFFF00', '#FF6600', '#FF0000'];

const logger = q.logger;


class RamUsage extends q.DesktopApp {
  constructor() {
    super();
    this.pollingInterval = 3000; // run() every 3000 ms
    logger.info("RAM Usage Meter ready to go!");
  }

  // call this function every pollingInterval
  async run() {
    return this.getRamUsage().then(percent => {
      // return signal each 3000ms, it depends of percent value
      return new q.Signal({
        points: [this.generateColor(percent)],
        name: "RAM Usage",
        message: "Memory used: " + Math.round(percent) + "%",
        isMuted: true, // don't flash the Q button on each signal
      });
    });
  }

  async getRamUsage() {
    return new Promise((resolve) => {
      si.mem(cb => {
       // get memory and create percent
       var percent = (cb.active/cb.total)*100;
       resolve(percent);
      });

 
    })
  }

  generateColor(percent) {

    let color =[];

    switch (true){

      case percent < 30:
        // return first color
        color.push(new q.Point(colors[0]));
        break;

      case percent < 50:
        // return second color
        color.push(new q.Point(colors[1]));
        break;

      case percent < 70:
        // return third color
        color.push(new q.Point(colors[2]));
        break;
      
      case percent < 85:
        // return fourth color
        color.push(new q.Point(colors[3]));
        break;

      case percent <= 100:
        // return fifth color
        color.push(new q.Point(colors[4]));
        break;

      default:
        // Something wrong happened, percent>100, return white color
        color.push(new q.Point("#FFFFFF"));
        break;

    };


    return color;
  }

}

module.exports = {
  RamUsage: RamUsage
};

const ramUsage = new RamUsage();
